source_manager.go

```diff
@@ -141,15 +158,121 @@ func NewSourceManager(an ProjectAnalyzer, cachedir string) (*SourceMgr, error) {
 		}
 	}
 
-	return &SourceMgr{
+	sm := &SourceMgr{
 		cachedir: cachedir,
 		lf:       fi,
 		srcs:     make(map[string]source),
 		srcfuts:  make(map[string]*unifiedFuture),
 		an:       an,
 		dxt:      pathDeducerTrie(),
 		rootxt:   newProjectRootTrie(),
-	}, nil
+		qch:      make(chan struct{}),
+	}
+
+	return sm, nil
+}
+
+// SetUpSigHandling sets up typical signal handling for a SourceMgr. It will
+// register a signal handler to be notified on:
+//
+//  - syscall.SIGINT
+//  - syscall.SIGHUP
+//  - syscall.SIGTERM
+//  - syscall.SIGQUIT
+//  - os.Interrupt
+func SetUpSigHandling(sm *SourceMgr) {
+	sigch := make(chan os.Signal)
```

Comment:
Sounds like you have it under control, note, each signal will be handled serially, unless you're then launching more goroutines yourself, but I don't know exactly which logic you're referring to, but I wouldn't *think* it'd be required (but again, I haven't closely followed the extra shutdown steps you need to do), but certainly a buffer should be no cost and improve reliability.